sim/shaman/enhancement/rotation.go

```diff
@@ -81,17 +90,30 @@ func (rotation *AdaptiveRotation) DoAction(enh *EnhancementShaman, sim *core.Sim
 				return
 			}
 		}
-	}
 
-	if enh.Talents.MaelstromWeapon > 0 && enh.MaelstromWeaponAura.GetStacks() >= 1 {
 		lbCastTime := enh.LightningBolt.DefaultCast.CastTime - (time.Millisecond * time.Duration(500*enh.MaelstromWeaponAura.GetStacks()))
 		lbCastTime = enh.ApplyCastSpeed(lbCastTime)
+
 		timeUntilSwing := enh.AutoAttacks.NextAttackAt() - sim.CurrentTime
 		if sim.CurrentTime > enh.AutoAttacks.NextAttackAt() {
 			timeUntilSwing = enh.AutoAttacks.MH.SwingDuration
 		}
+
+		swingSpeed := core.MinDuration(enh.AutoAttacks.MainhandSwingSpeed(), enh.AutoAttacks.OffhandSwingSpeed())
+		swingSpeed = core.MaxDuration(swingSpeed, timeUntilSwing)
+		latency := time.Duration(enh.WeaveLatency) * time.Millisecond
+		swingSpeed -= latency
+
+		if swingSpeed < timeUntilSwing {
+			latency = timeUntilSwing - swingSpeed
+			timeUntilSwing -= latency
+		}
+
 		if lbCastTime < timeUntilSwing {
-			if !enh.LightningBolt.Cast(sim, target) {
+			if latency > 0 {
+				enh.isWeaving = true
+				enh.WaitUntil(sim, sim.CurrentTime+latency)
+			} else if !enh.LightningBolt.Cast(sim, target) {
```

Comment:
i think the entire weaving logic is gonna need a complete restructure

Reply:
this isnt like, a criticism of your code though its a criticism of mine lol

Reply:
yeah it bounces around to much, deff needs a restructure haha.